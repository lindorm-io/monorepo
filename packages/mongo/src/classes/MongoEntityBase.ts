import { IMongoEntity } from "../interfaces";

export class MongoEntityBase implements IMongoEntity {
  public readonly id!: string;
  public readonly rev!: number;
  public readonly seq!: number;
  public readonly createdAt!: Date;
  public readonly deletedAt!: Date | null;
  public readonly updatedAt!: Date;

  public expiresAt!: Date | null;
}
