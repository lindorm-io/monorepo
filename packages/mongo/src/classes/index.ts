export * from "./MongoBucket";
export * from "./MongoFileBase";
export * from "./MongoRepository";
export * from "./MongoSource";
