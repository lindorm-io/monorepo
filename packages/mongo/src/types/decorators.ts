export type FileDecoratorOptions = {
  chunkSizeBytes?: number;
  database?: string;
  name?: string;
  namespace?: string;
};
