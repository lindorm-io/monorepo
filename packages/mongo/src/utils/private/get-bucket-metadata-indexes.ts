import { EntityMetadata, MetaIndex } from "@lindorm/entity";

export const getBucketMetadataIndexes = (
  metadata: EntityMetadata,
): Array<Omit<MetaIndex, "target">> =>
  metadata.indexes.map((i) => ({
    ...i,
    index: i.keys.reduce(
      (acc, { key, direction }) => ({
        ...acc,
        [`metadata.${key}`]: direction,
      }),
      {},
    ),
  }));
