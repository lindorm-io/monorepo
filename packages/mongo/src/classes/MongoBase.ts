import { ILogger } from "@lindorm/logger";
import { Collection, Db, Document, IndexSpecification, MongoClient } from "mongodb";
import { MongoRepositoryError } from "../errors";
import { MongoBaseIndex, MongoBaseOptions, MongoIndexOptions } from "../types";

export abstract class MongoBase<D extends Document> {
  protected readonly client: MongoClient;
  protected readonly collectionName: string;
  protected readonly databaseName: string;
  protected readonly indexes: Array<MongoIndexOptions<D>>;
  protected readonly logger: ILogger;

  protected _collection: Collection<D> | undefined;
  protected _database: Db | undefined;

  protected constructor(options: MongoBaseOptions<D>) {
    this.logger = options.logger.child(["MongoBase"]);

    this.collectionName = options.collectionName;
    this.databaseName = options.databaseName;
    this.indexes = options.indexes;
    this.client = options.client;
  }

  // getters

  protected get collection(): Collection<any> {
    if (!this._collection) {
      this._collection = this.database.collection(this.collectionName);
    }
    return this._collection;
  }

  protected get database(): Db {
    if (!this._database) {
      this._database = this.client.db(this.databaseName);
    }
    return this._database;
  }

  // public

  public async setup(): Promise<void> {
    const indexes: Array<MongoBaseIndex> = this.indexes.map((item) => ({
      index: item.index as IndexSpecification,
      options: {
        unique: item.unique ?? false,
        ...(item.name ? { name: item.name } : {}),
        ...(item.nullable?.length
          ? {
              partialFilterExpression: {
                ...item.nullable.reduce(
                  (obj, key) => ({ ...obj, [key]: { $exists: true } }),
                  {},
                ),
                ...item.finite?.reduce((obj, key) => ({ ...obj, [key]: { $gt: 0 } }), {}),
              },
            }
          : {}),
        ...(item.options ?? {}),
      },
    }));

    for (const { index, options } of indexes) {
      this.logger.debug("Creating index", { index, options });

      try {
        const result = await this.collection.createIndex(index, options);

        this.logger.silly("Index created", { result });
      } catch (error: any) {
        this.logger.silly("Mongo setup error", error);

        throw new MongoRepositoryError(error.message, { error });
      }
    }
  }
}
