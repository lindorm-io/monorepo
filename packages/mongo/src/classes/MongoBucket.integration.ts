import { createMockLogger } from "@lindorm/logger";
import { randomUUID } from "crypto";
import { createReadStream } from "fs";
import MockDate from "mockdate";
import { MongoClient } from "mongodb";
import { join } from "path";
import { Readable } from "stream";
import { TestBucket } from "../__fixtures__/test-bucket";
import { TestFile } from "../__fixtures__/test-file";
import { getReadStreamContent } from "../__fixtures__/utils";
import { MongoBucketError } from "../errors";
import { FileUpload } from "../types";

const MockedDate = new Date("2024-01-01T08:00:00.000Z");
MockDate.set(MockedDate);

describe("MongoBucket", () => {
  let client: MongoClient;
  let bucket: TestBucket;
  let stream: () => Readable;
  let metadata: FileUpload<TestFile>;

  beforeAll(async () => {
    client = new MongoClient("mongodb://root:example@localhost/admin?authSource=admin");
    await client.connect();
    bucket = new TestBucket(client, createMockLogger());
  });

  beforeEach(async () => {
    stream = () =>
      createReadStream(join(__dirname, "..", "__fixtures__", "test-file-upload.txt"));

    metadata = {
      encoding: "7bit",
      extraOne: 1,
      extraTwo: 2,
      hash: "hash",
      hashAlgorithm: "sha256",
      mimeType: "text/plain",
      name: randomUUID(),
      originalName: "test-file-upload.txt",
      size: 18,
      strategy: "local",
    };
  });

  afterAll(async () => {
    await client.close();
  });

  test("should setup", async () => {
    await expect(bucket.setup()).resolves.not.toThrow();
  }, 30000);

  test("should delete many files", async () => {
    const f1 = await bucket.upload(stream(), metadata);
    const f2 = await bucket.upload(stream(), metadata);

    await expect(bucket.delete({ name: f1.name })).resolves.not.toThrow();

    await expect(bucket.findOne({ filename: f1.filename })).resolves.toBeNull();
    await expect(bucket.download(f1.filename)).rejects.toThrow();

    await expect(bucket.findOne({ filename: f2.filename })).resolves.toBeNull();
    await expect(bucket.download(f2.filename)).rejects.toThrow();
  });

  test("should download file by name", async () => {
    const result = await bucket.upload(stream(), metadata);
    const download = await bucket.download(result.filename);

    expect(download.chunkSize).toEqual(261120);
    expect(download.filename).toEqual(result.filename);
    expect(download.length).toEqual(18);
    expect(download.mimeType).toEqual("text/plain");
    expect(download.originalName).toEqual("test-file-upload.txt");
    expect(download.uploadDate).toEqual(MockedDate);

    await expect(getReadStreamContent(download.stream)).resolves.toEqual(
      "test file content\n",
    );
  });

  test("should find many files", async () => {
    const f1 = await bucket.upload(stream(), metadata);
    const f2 = await bucket.upload(stream(), metadata);

    await expect(bucket.find({ name: f1.name })).resolves.toEqual([f1, f2]);
    await expect(bucket.find({ name: randomUUID() })).resolves.toEqual([]);
  });

  test("should find one file", async () => {
    const file = await bucket.upload(stream(), metadata);

    await expect(bucket.findOne({ name: file.name })).resolves.toEqual(file);
    await expect(bucket.findOne({ name: randomUUID() })).resolves.toBeNull();
  });

  test("should find one or fail", async () => {
    const file = await bucket.upload(stream(), metadata);

    await expect(bucket.findOneOrFail({ name: file.name })).resolves.toEqual(file);
    await expect(bucket.findOneOrFail({ name: randomUUID() })).rejects.toThrow(
      MongoBucketError,
    );
  });

  test("should find one file by filename", async () => {
    const file = await bucket.upload(stream(), metadata);

    await expect(bucket.findOne({ filename: file.filename })).resolves.toEqual(file);
    await expect(bucket.findOne({ filename: randomUUID() })).resolves.toBeNull();
  });

  test("should find one file by filename or fail", async () => {
    const file = await bucket.upload(stream(), metadata);

    await expect(bucket.findOneOrFail({ filename: file.filename })).resolves.toEqual(
      file,
    );
    await expect(bucket.findOneOrFail({ filename: randomUUID() })).rejects.toThrow(
      MongoBucketError,
    );
  });

  test("should upload file", async () => {
    await expect(bucket.upload(stream(), metadata)).resolves.toEqual({
      chunkSize: 261120,
      encoding: "7bit",
      extraOne: 1,
      extraTwo: 2,
      filename: expect.any(String),
      hash: "hash",
      hashAlgorithm: "sha256",
      length: 18,
      mimeType: "text/plain",
      name: metadata.name,
      originalName: "test-file-upload.txt",
      size: 18,
      strategy: "local",
      uploadDate: MockedDate,
    });
  });
});
