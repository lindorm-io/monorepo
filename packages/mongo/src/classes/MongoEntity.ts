import { EntityBase } from "@lindorm/entity";
import { IMongoEntity } from "../interfaces";
import { MongoEntityConfig } from "../types";

export class MongoEntity extends EntityBase implements IMongoEntity {
  public readonly rev!: number;
  public readonly seq!: number;
  public readonly deletedAt!: Date | null;
  public expiresAt!: Date | null;
}

export const MONGO_ENTITY_CONFIG: MongoEntityConfig<MongoEntity> = {
  revisionAttribute: "rev",
  sequenceAttribute: "seq",
  deleteAttribute: "deletedAt",
  ttlAttribute: "expiresAt",
};
