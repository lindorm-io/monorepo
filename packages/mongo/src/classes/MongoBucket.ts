import { snakeCase } from "@lindorm/case";
import { ILogger } from "@lindorm/logger";
import { Constructor, Dict } from "@lindorm/types";
import { randomUUID } from "crypto";
import { DeleteOptions, Filter, FindOptions, GridFSBucket } from "mongodb";
import { Readable } from "stream";
import { MongoBucketError } from "../errors";
import { IMongoBucket, IMongoFile } from "../interfaces";
import { GridFSDocument } from "../interfaces/private";
import {
  FileDownload,
  FileIndex,
  FileMetadata,
  MongoBucketOptions,
  ValidateMongoFileFn,
} from "../types";
import { MongoBase } from "./MongoBase";

export class MongoBucket<F extends IMongoFile>
  extends MongoBase<FileIndex<F>>
  implements IMongoBucket<F>
{
  private readonly FileConstructor: Constructor<F>;
  private readonly bucketName: string;
  private readonly chunkSizeBytes: number | undefined;
  private readonly validate: ValidateMongoFileFn<F> | undefined;
  private _grid: GridFSBucket | undefined;

  protected readonly logger: ILogger;

  public constructor(options: MongoBucketOptions<F>) {
    super({
      client: options.client,
      collectionName: MongoBucket.createCollectionName(options),
      databaseName: options.database,
      logger: options.logger,
      indexes: [
        {
          index: { filename: 1 },
          unique: true,
        },
        {
          index: { "metadata.mimeType": 1 },
          unique: false,
        },
        {
          index: { "metadata.originalName": 1 },
          unique: false,
          nullable: ["metadata.originalName"],
        },
        ...(options.indexes
          ? options.indexes.map((item) => ({
              ...item,
              index: Object.entries(item.index).reduce(
                (acc, [key, value]) => ({
                  ...acc,
                  [`metadata.${key}`]: value,
                }),
                {},
              ),
              ...(item.nullable
                ? { nullable: item.nullable.map((key) => `metadata.${key as string}`) }
                : {}),
            }))
          : []),
      ],
    });

    this.logger = options.logger.child(["MongoMongoBucket", options.File.name]);

    this.FileConstructor = options.File;
    this.bucketName = MongoBucket.createBucketName(options);
    this.chunkSizeBytes = options.chunkSizeBytes;
    this.validate = options.validate;
  }

  // getters

  protected get grid(): GridFSBucket {
    if (!this._grid) {
      this._grid = new GridFSBucket(this.client.db(this.databaseName), {
        bucketName: this.bucketName,
        ...(this.chunkSizeBytes ? { chunkSizeBytes: this.chunkSizeBytes } : {}),
      });
    }

    return this._grid;
  }

  // public

  public async delete(criteria: Filter<F>, options?: DeleteOptions): Promise<void> {
    const start = Date.now();

    try {
      const filter = this.parseFilter(criteria);
      const result = await this.collection.find(filter).toArray();

      for (const document of result) {
        await this.grid.delete(document._id);
      }

      await this.collection.deleteMany(filter, options);

      this.logger.debug("Bucket done: delete", {
        input: {
          filter,
        },
        result: {
          ...result,
          time: Date.now() - start,
        },
      });
    } catch (error: any) {
      this.logger.error("MongoBucket error", error);
      throw new MongoBucketError("Unable to delete files", { error });
    }
  }

  public async deleteByFilename(filename: string): Promise<void> {
    const start = Date.now();

    try {
      const result = await this.collection.findOne({ filename });

      await this.collection.deleteOne({ _id: result._id });
      await this.grid.delete(result._id);

      this.logger.debug("Bucket done: deleteByFilename", {
        input: {
          filename,
        },
        result: {
          ...result,
          time: Date.now() - start,
        },
      });
    } catch (error: any) {
      this.logger.error("MongoBucket error", error);
      throw new MongoBucketError("Unable to delete file", { error });
    }
  }

  public async download(filename: string): Promise<FileDownload> {
    try {
      const file = await this.findOneByFilenameOrFail(filename);
      const stream = this.grid.openDownloadStreamByName(filename);

      return { ...file, stream };
    } catch (error: any) {
      this.logger.error("MongoBucket error", error);
      throw new MongoBucketError("Error downloading file", {
        data: { filename },
        error,
      });
    }
  }

  public async find(criteria?: Filter<F>, options?: FindOptions<F>): Promise<Array<F>> {
    const start = Date.now();

    try {
      const filter = this.parseFilter(criteria);
      const result = await this.collection.find(filter, options).toArray();

      this.logger.debug("Bucket done: find", {
        input: {
          criteria,
          filter,
          options,
        },
        result: {
          count: result.length,
          time: Date.now() - start,
        },
      });

      return result.map((document) => this.create(document));
    } catch (error: any) {
      this.logger.error("MongoBucket error", error);
      throw new MongoBucketError("Unable to find files", { error });
    }
  }

  public async findOne(criteria: Filter<F>, options?: FindOptions<F>): Promise<F | null> {
    const start = Date.now();

    try {
      const filter = this.parseFilter(criteria);
      const document = await this.collection.findOne(filter, options);

      this.logger.debug("Bucket done: findOne", {
        input: {
          criteria,
          filter,
          options,
        },
        result: {
          document,
          time: Date.now() - start,
        },
      });

      if (!document) return null;

      return this.create(document);
    } catch (error: any) {
      this.logger.error("MongoBucket error", error);
      throw new MongoBucketError("Unable to find file", { error });
    }
  }

  public async findOneOrFail(criteria: Filter<F>, options?: FindOptions<F>): Promise<F> {
    const document = await this.findOne(criteria, options);

    if (!document) {
      throw new MongoBucketError("File not found", { debug: { criteria, options } });
    }

    return document;
  }

  public async findOneByFilename(filename: string): Promise<F | null> {
    return this.findOne({ filename } as any);
  }

  public async findOneByFilenameOrFail(filename: string): Promise<F> {
    return this.findOneOrFail({ filename } as any);
  }

  public async upload(stream: Readable, metadata: FileMetadata<F>): Promise<F> {
    this.logger.debug("Uploading file", { metadata });

    this.validateMetadata(metadata);

    try {
      const filename = randomUUID();
      const upload = this.grid.openUploadStream(filename, { metadata });

      stream.pipe(upload);

      this.logger.debug("File upload stream initialised", { filename });

      await new Promise<void>((resolve, reject) => {
        upload.once("finish", resolve);
        upload.once("error", reject);
      });

      const { done, state, gridFSFile } = upload;

      if (!done) {
        throw new Error("File upload not done");
      }

      if (!gridFSFile) {
        throw new Error("GridFSFile not initialized");
      }

      const { length, chunkSize, uploadDate } = gridFSFile;

      this.logger.error("Bucket done: upload", {
        chunkSize,
        done,
        length,
        metadata,
        state,
        uploadDate,
      });

      return await this.findOneByFilenameOrFail(filename);
    } catch (error: any) {
      this.logger.error("Error uploading file", error);

      throw new MongoBucketError("Error uploading file", { error });
    }
  }

  // private

  private create(document: GridFSDocument): F {
    const file = new this.FileConstructor(document.metadata);

    file.chunkSize = document.chunkSize;
    file.filename = document.filename;
    file.length = document.length;
    file.mimeType = document.metadata.mimeType;
    file.originalName = document.metadata.originalName;
    file.uploadDate = document.uploadDate;

    this.logger.debug("Created file", { file });

    return file;
  }

  private parseFilter(criteria: Filter<any> = {}): Filter<any> {
    const parsed: Dict = {};

    for (const [key, value] of Object.entries(criteria)) {
      if (
        key === "chunkSize" ||
        key === "filename" ||
        key === "length" ||
        key === "uploadDate"
      ) {
        parsed[key] = value;
        continue;
      }

      parsed[`metadata.${key}`] = value;
    }

    this.logger.debug("Parsed criteria", { criteria, parsed });

    return parsed;
  }

  private validateMetadata(metadata: FileMetadata<F>): void {
    if (!this.validate) return;

    this.validate(metadata);
  }

  // private static

  private static createBucketName<F extends IMongoFile>(
    options: MongoBucketOptions<F>,
  ): string {
    const nsp = options.namespace ? `${snakeCase(options.namespace)}_` : "";
    const name = snakeCase(options.File.name);

    return `${nsp}${name}`;
  }

  private static createCollectionName<F extends IMongoFile>(
    options: MongoBucketOptions<F>,
  ): string {
    return `${MongoBucket.createBucketName(options)}.files`;
  }
}
