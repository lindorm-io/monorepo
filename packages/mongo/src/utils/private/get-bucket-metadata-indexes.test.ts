import { globalEntityMetadata } from "@lindorm/entity";
import { MongoFileBase } from "../../classes";
import { File } from "../../decorators";
import { getBucketMetadataIndexes } from "./get-bucket-metadata-indexes";

describe("getBucketMetadataIndexes", () => {
  test("should return array for MongoFileBase", () => {
    @File()
    class TestMongoFileBase extends MongoFileBase {}

    const metadata = globalEntityMetadata.get(TestMongoFileBase);

    expect(getBucketMetadataIndexes(metadata)).toMatchSnapshot();
  });
});
