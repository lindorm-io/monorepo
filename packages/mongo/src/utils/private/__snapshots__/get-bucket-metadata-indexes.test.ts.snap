// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBucketMetadataIndexes should return array for MongoFileBase 1`] = `
[
  {
    "index": {
      "metadata.mimeType": "asc",
    },
    "keys": [
      {
        "direction": "asc",
        "key": "mimeType",
      },
    ],
    "name": null,
    "options": {},
    "unique": false,
  },
  {
    "index": {
      "metadata.originalName": "asc",
    },
    "keys": [
      {
        "direction": "asc",
        "key": "originalName",
      },
    ],
    "name": null,
    "options": {},
    "unique": false,
  },
]
`;
